### PR TITLE
ovl_En_Part

### PR DESCRIPTION
**Pull request by maekclena**

### CODE REVIEW

File: src/overlays/actors/ovl_En_Part/z_en_part.c
Comment (Dragorn421) [inline src/overlays/actors/ovl_En_Part/z_en_part.c:49]: Given their usage I would think those are textures (whicharrays of UNK_TYPE is better for I think) but I have no idea what object this ends up using so I can't check

(this actor must be the most janky for its object management, which comes from the spawn z rotation in a hardcoded if case in Actor_Spawn)

File: src/overlays/actors/ovl_En_Part/z_en_part.c
Comment (Dragorn421) [inline src/overlays/actors/ovl_En_Part/z_en_part.c:140]: what about `&&`ing these?

File: src/overlays/actors/ovl_En_Part/z_en_part.c
Comment (Dragorn421) [inline src/overlays/actors/ovl_En_Part/z_en_part.c:199]: `this->delay--;` ?

File: src/overlays/actors/ovl_En_Part/z_en_part.c
Comment (Dragorn421) [inline src/overlays/actors/ovl_En_Part/z_en_part.c:211]: What about `<= -40` instead of `< -39`

**Review by Dragorn421 — CHANGES_REQUESTED** [review]
well what this actor does is very mysterious to me, looks like an effect but it's not?

**Comment by maekclena** [comment]
This actor handles colored flames when enemies die, stalfos' bones flying and probably other similar death effects.

File: src/code/z_actor.c
Comment (Dragorn421) [inline src/code/z_actor.c:3123]: You can also change `arg1->unk_0C` to be of type `Gfx**`
(in `struct_80032E24`)

File: src/overlays/actors/ovl_En_Part/z_en_part.c
Comment (Dragorn421) [inline src/overlays/actors/ovl_En_Part/z_en_part.c:297]: based on this, I would rename `rotation` to `rotationZ` or `rotZ` for shorts

File: src/overlays/actors/ovl_En_Part/z_en_part.c
Comment (Dragorn421) [inline src/overlays/actors/ovl_En_Part/z_en_part.c:226]: I would rename `dist` to `diffsSum` or something like that because `Math_SmoothScaleMaxMinF` returns the absolute difference between the computed value and the target, and `dist` is summing position differences with a rotation difference

File: src/code/z_actor.c
Comment (Dragorn421) [inline src/code/z_actor.c:4088]: may be neat to create a struct for `arg3`, or just make it a `f32` array, I don't think `Vec3f` is appropriate now that we know what it's used for

You can also update `func_800358DC`'s prototype (types and names of arguments) according to your findings

File: src/overlays/actors/ovl_En_Part/z_en_part.h
Comment (Dragorn421) [inline src/overlays/actors/ovl_En_Part/z_en_part.h:14]: I think this actor uses `delay` like "when `delay` reaches 0, stop doing things"
I'm not sure I like "delay" because that often implies "wait to do something" rather than "do something until"

For one of the actions (4, `func_80ACE7E8`) "stop doing things" doesn't mean "remove the actor" like for the others (it still draws), so `lifetime` isn't a good name
I'm not very inspired at the moment

**Review by Dragorn421 — CHANGES_REQUESTED** [review]

File: src/overlays/actors/ovl_En_Part/z_en_part.h
Comment (maekclena) [reply src/overlays/actors/ovl_En_Part/z_en_part.h:14]: From my tests, delay was used as "wait until spawning effect". See also `func_80ACE13C` where all the function calls are effects.

Comment (Dragorn421) [reply src/overlays/actors/ovl_En_Part/z_en_part.h:14]: But doesn't `EnPart_Draw` get called meanwhile? It draws `this->displayList`

Comment (maekclena) [reply src/overlays/actors/ovl_En_Part/z_en_part.h:14]: I tested and it seems, at least for stalchild, that `EnPart_Draw` draws the bones. Since `delay` only matter for the flames, a good name might be `effectDelay`.

**Review by Dragorn421 — APPROVED** [review]

File: src/overlays/actors/ovl_En_Part/z_en_part.c
Comment (Roman971) [inline src/overlays/actors/ovl_En_Part/z_en_part.c:117]:
```suggestion
    s32 i;
    Vec3f pos;
    Vec3f velocity = { 0.0f, 0.0f, 0.0f };
    Vec3f accel = { 0.0f, 0.0f, 0.0f };
    Vec3f zero = { 0.0f, 0.0f, 0.0f };
```

File: src/overlays/actors/ovl_En_Part/z_en_part.c
Comment (Roman971) [inline src/overlays/actors/ovl_En_Part/z_en_part.c:225]:
```suggestion
    Vec3f zero = { 0.0f, 0.0f, 0.0f };;
```

File: src/overlays/actors/ovl_En_Part/z_en_part.c
Comment (Roman971) [inline src/overlays/actors/ovl_En_Part/z_en_part.c:197]: If you use an else here (or just a new block), you can inline the vector definitions so you don't need to copy static variables (see my previous comments)

File: src/overlays/actors/ovl_En_Part/z_en_part.c
Comment (Roman971) [inline src/overlays/actors/ovl_En_Part/z_en_part.c:258]: I would personally leave this array of functions outside the file (right above it), but if you want to keep it inside the function I suggest formatting it like this (note the trailing comma):
```suggestion
    static EnPartActionFunc sActionFuncs[] = {
        func_80ACDDE8, func_80ACE13C, func_80ACE5B8, func_80ACE5C8, func_80ACE7E8,
    };
    EnPart* this = THIS;
```

File: src/overlays/actors/ovl_En_Part/z_en_part.h
Comment (Roman971) [inline src/overlays/actors/ovl_En_Part/z_en_part.h:14]: I could be wrong but I think this is just the timer for the whole actor
```suggestion
    /* 0x14E */ s16 timer;
```

File: src/overlays/actors/ovl_En_Part/z_en_part.c
Comment (Roman971) [inline src/overlays/actors/ovl_En_Part/z_en_part.c:200]: Considering the usage, I suggest naming this "prev" or "old"
```suggestion
        prevInvincibilityTimer = player->invincibilityTimer;
```

**Review by Roman971 — COMMENTED** [review]
Good job with this actor. I just have a few possible improvements

**Review by Roman971 — APPROVED** [review]

**Comment by fig02** [comment]
After you merge effects in, theres a good amount of args that will make more sense in decimal. Most of them are named so it will be clear what they should be. Ill give a more in-depth review after that :D

File: src/overlays/actors/ovl_En_Part/z_en_part.c
Comment (fig02) [inline src/overlays/actors/ovl_En_Part/z_en_part.c:152]: randXZ (the last arg there) is a bool so you can make that `true`
I also prefer to call `zero` `zeroVec` but I leave that up to you

File: src/overlays/actors/ovl_En_Part/z_en_part.c
Comment (fig02) [inline src/overlays/actors/ovl_En_Part/z_en_part.c:270]: we've started going with dList and dList head as the abbreviations for these, but we havent made the change project wide yet. dont have to change it now if you dont want, but just letting you know

File: src/overlays/actors/ovl_En_Part/z_en_part.c
Comment (fig02) [inline src/overlays/actors/ovl_En_Part/z_en_part.c:4]: I prefer the term "effects" over particles, cause some of the effects do stuff like making sounds. and other things that arent really particles.
I would say Effect Spawner instead of Particle Emitter but thats just me :D

**Review by fig02 — APPROVED** [review]

**Pull request merged**
